### PR TITLE
Use vc-status-list-context@3.0 

### PR DESCRIPTION
**Pull request by JSAssassin**

### CODE REVIEW

File: main.js [hunk 1]
```diff
@@ -81,11 +81,10 @@ export function statusTypeMatches({credential} = {}) {
     return false;
   }
   const credentialStatuses = _getStatuses({credential});
-  // at least one "credentialStatus.type" must match
-  // RevocationList2021Status or SuspensionList2021Status
+  // at least o"credentialStatus.type" must match
```
Comment (dlongley) [inline main.js:84]: ^please fix whatever happened here. :)

File: main.js [hunk 2]
```diff
@@ -140,8 +132,9 @@ export function getCredentialStatus({credential, statusPurpose} = {}) {
   const credentialStatuses = _getStatuses({credential});
   const result = credentialStatuses.filter(
     credentialStatus => _validateStatus({credentialStatus})).find(
-    // check for matching `type` (legacy) or `statusPurpose`
-    cs => cs.type === expectedType || cs.statusPurpose === statusPurpose);
+    // check for matching `type` or `statusPurpose`
+    cs => cs.type === 'StatusList2021Entry' ||
+      cs.statusPurpose === statusPurpose);
```
Comment (dlongley) [inline main.js:137]: Why is this still an `or` condition now that the legacy has been removed? Seems like this logic needs to be made consistent with the data format and how this particular function works.

File: tests/30-main.spec.js [hunk 3]
```diff
@@ -700,11 +695,28 @@ describe('main', () => {
     err.message.should.contain('@context" must include');
   });
 
-  it('should fail when credentialStatus is not an object for ' +
+  it('should fail when "credentialStatus" is not an object for ' +
     '"getCredentialStatus"', async () => {
-    const id = 'https://example.com/status/1';
-    const list = await createList({length: 100000});
-    const credential = await createCredential({id, list});
+    const credential = {
```
Comment (dlongley) [inline tests/30-main.spec.js:700]: Why was this changed to be hard-coded vs. using the `createCredential` function?

**Review by dlongley — CHANGES_REQUESTED** [review]

File: main.js (hunk 1 above)
Comment (JSAssassin) [reply main.js]: Fixed [5bc424e](https://github.com/digitalbazaar/vc-status-list/pull/6/commits/5bc424ee4a3e586e5374ea26430d7c045b83cc4b)

File: main.js (hunk 2 above)
Comment (JSAssassin) [reply main.js]: Updated [11db1a3](https://github.com/digitalbazaar/vc-status-list/pull/6/commits/11db1a35e3a6c9aa948dc63b06d6159692cd79b4)

File: tests/30-main.spec.js (hunk 3 above)
Comment (JSAssassin) [reply tests/30-main.spec.js:700]: This was changed to be hard-coded since `createCredential()` doesn't return a credential with "credentialStatus" property on it.

File: main.js [hunk 4]
```diff
@@ -140,8 +131,8 @@ export function getCredentialStatus({credential, statusPurpose} = {}) {
   const credentialStatuses = _getStatuses({credential});
   const result = credentialStatuses.filter(
     credentialStatus => _validateStatus({credentialStatus})).find(
-    // check for matching `type` (legacy) or `statusPurpose`
-    cs => cs.type === expectedType || cs.statusPurpose === statusPurpose);
+    // check for matching `statusPurpose`
+    cs => cs.statusPurpose === statusPurpose);
```
Comment (dlongley) [inline main.js:135]: Is this the right resolution to the problem from before? What happens if the type doesn't match `StatusList2021Entry` (assuming that's the right type here)? Is this checked another way? It seems like maybe we're missing test coverage for this.

File: main.js [hunk 5]
```diff
@@ -140,8 +131,8 @@ export function getCredentialStatus({credential, statusPurpose} = {}) {
...
   if(!result) {
     throw new Error(
       `credentialStatus with status purpose "${statusPurpose}" not found.`);
```
Comment (dlongley) [inline main.js:138]:
```suggestion
      `"credentialStatus" with status purpose "${statusPurpose}" not found.`);
```

File: main.js [hunk 6]
```diff
@@ -222,8 +213,8 @@ async function _checkStatus({
   // get JSON RevocationList
```
Comment (dlongley) [inline main.js:213]: There's revocation list language in the comments here -- is that still proper?

File: main.js [hunk 7]
```diff
@@ -222,8 +213,8 @@ async function _checkStatus({
...
   const {credentialSubject: rl} = slCredential;
 
-  if(rl.type !== 'RevocationList2021') {
-    throw new Error('Revocation list type must be "RevocationList2021".');
+  if(rl.type !== 'StatusList2021') {
+    throw new Error('Revocation list type must be "StatusList2021".');
```
Comment (dlongley) [inline main.js:217]: Is using revocation list language here still proper?

File: main.js [hunk 8]
```diff
@@ -222,8 +213,8 @@ async function _checkStatus({
...
   }
 
   // decode list from RL VC
```
Comment (dlongley) [inline main.js:220]: Should this be an `SL VC` instead? Can you check the rest of this file to make sure the right language is used?

File: tests/30-main.spec.js (hunk 3 above)
Comment (dlongley) [reply tests/30-main.spec.js:700]: But it gets deleted below anyway (line 723) ... right? Isn't the point of this test to check for when it's not present?

**Review by dlongley — CHANGES_REQUESTED** [review]

File: main.js (hunk 7 above)
Comment (aljones15) [reply main.js:217]: @dlongley just so you know some of the issues you have in this PR are from the multiple statuses PR I made so my apologies to @JSAssassin for mistakes I made there.

Comment (dlongley) [reply main.js:217]: No spanish inquisition here :P ... just noticing things we need fixed.

File: main.js (hunk 6 above)
Comment (JSAssassin) [reply main.js]: Removed [b9a53d6](https://github.com/digitalbazaar/vc-status-list/pull/6/commits/b9a53d64e2c4ae814291ee6aae9a9cd26d3758c6)

File: tests/30-main.spec.js (hunk 3 above)
Comment (JSAssassin) [reply tests/30-main.spec.js]: Yeah, makes sense. I've reverted it back to how it was previously. [e7c328b](https://github.com/digitalbazaar/vc-status-list/pull/6/commits/e7c328bbaaf807a9ac9add67438aa6d8f8a06d70)

File: main.js (hunk 7 above)
Comment (JSAssassin) [reply main.js]: Updated [b9a53d6](https://github.com/digitalbazaar/vc-status-list/pull/6/commits/b9a53d64e2c4ae814291ee6aae9a9cd26d3758c6)

**Review by dlongley — CHANGES_REQUESTED** [review]
See comments.

File: main.js (hunk 4 above)
Comment (JSAssassin) [reply main.js:138]: @dlongley can you pls say more about what the expected behavior for that above case should be and also, is `statusPurpose` a required property of `credentialStatus`?

Comment (dlongley) [reply main.js:138]: `statusPurpose` is only a required property of `credentialStatus` when the credential status type is `StatusList2021Entry`. A VC may have multiple credential status entries, however, some that may not have that type. We want to ignore those. We should not be throwing on unknown status types. Does that explain everything?

Comment (JSAssassin) [reply main.js:138]: @dlongley yeah, makes sense. thanks

Comment (JSAssassin) [reply main.js:137]: Updated and added test here [946bb2e](https://github.com/digitalbazaar/vc-status-list/pull/6/commits/946bb2ef6996b1da6750340f5f4acfd0f2523431).

File: main.js [hunk 9]
```diff
@@ -337,14 +338,18 @@ function _isObject({credential}) {
...
- * @returns {Array<object>} An array of statuses.
+ * @returns {Array<object>} An array of statuses with type
+ *   "StatusList2021Entry".
  */
 function _getStatuses({credential}) {
   const {credentialStatus} = credential;
-  if(Array.isArray(credentialStatus)) {
-    return credentialStatus;
+  if(Array.isArray(credentialStatus) && credentialStatus.length > 0) {
+    return credentialStatus.filter(cs => cs.type === 'StatusList2021Entry');
```
Comment (dlongley) [inline main.js:347]: This can return an empty array -- and then the checks above that are supposed to throw when there are no matching types won't throw. Looks like we need another test with an empty `credentialStatus` array on a VC.

**Review by dlongley — CHANGES_REQUESTED** [review]

File: main.js (hunk 9 above)
Comment (JSAssassin) [reply main.js]: Fixed and added tests here [53a86b2](https://github.com/digitalbazaar/vc-status-list/pull/6/commits/53a86b20e3e57c87d9239f81d97953a9897b39e3)

File: main.js [hunk 10]
```diff
@@ -340,18 +338,26 @@ function _isObject({credential}) {
...
+    if(statuses.length !== 0) {
+      return statuses;
+    }
+    return null;
+  }
+  if(credentialStatus && credentialStatus.type === 'StatusList2021Entry') {
+    return [credentialStatus];
   }
-  return [credentialStatus];
+  return null;
```
Comment (dlongley) [inline main.js:357]: Upon further reflection, what does returning `null` instead of an empty array buy us here? Wouldn't it be simpler if the return value of this method was always an array -- and if callers can if it's empty or not they should just check `.length` on that value?

Comment (JSAssassin) [reply main.js]: Updated to return empty array instead of null here [5c0e9df](https://github.com/digitalbazaar/vc-status-list/pull/6/commits/5c0e9df1cc54c8b1dbb7683c78a19aa5debe54e7)

File: main.js [hunk 11]
```diff
@@ -81,16 +81,10 @@ export function statusTypeMatches({credential} = {}) {
...
-      credentialStatus.type === 'SuspensionList2021Status')) {
-      // status type does not match
-      return false;
-    }
-    return true;
-  });
+  if(credentialStatuses.length === 0) {
+    return false;
+  }
+  return true;
```
Comment (dlongley) [inline main.js:87]:
```suggestion
  return credentialStatuses.length > 0;
```

File: main.js [hunk 12]
```diff
@@ -125,26 +119,25 @@ export function assertStatusList2021Context({credential} = {}) {
...
   assertStatusList2021Context({credential});
-  // older revisions of `StatusList2021` used different types to express
-  // the status purpose:
-  const statusPurposes = {
-    revocation: 'RevocationList2021Status',
-    suspension: 'SuspensionList2021Status'
-  };
-  const expectedType = statusPurposes[statusPurpose];
+  if(!(statusPurpose && typeof statusPurpose === 'string')) {
+    throw new Error('"statusPurpose" string is required.');
```
Comment (dlongley) [inline main.js:123]:
```suggestion
    throw new TypeError('"statusPurpose" must be a string.');
```

File: main.js [hunk 13]
```diff
@@ -125,26 +119,25 @@ export function assertStatusList2021Context({credential} = {}) {
...
+  }
   // get and validate status
   if(!(credential.credentialStatus &&
     typeof credential.credentialStatus === 'object')) {
     throw new Error('"credentialStatus" is missing or invalid.');
   }
   const credentialStatuses = _getStatuses({credential});
+  if(credentialStatuses.length === 0) {
+    throw new Error('"credentialStatus.type" must be "StatusList2021Entry".');
+  }
```
Comment (dlongley) [inline main.js:133]:
```suggestion
```

File: main.js [hunk 14]
```diff
@@ -125,26 +119,25 @@ export function assertStatusList2021Context({credential} = {}) {
...
   const result = credentialStatuses.filter(
     credentialStatus => _validateStatus({credentialStatus})).find(
-    // check for matching `type` (legacy) or `statusPurpose`
-    cs => cs.type === expectedType || cs.statusPurpose === statusPurpose);
+    // check for matching `statusPurpose`
+    cs => cs.statusPurpose === statusPurpose);
   if(!result) {
     throw new Error(
-      `credentialStatus with status purpose "${statusPurpose}" not found.`);
+      `"credentialStatus" with status purpose "${statusPurpose}" not found.`);
```
Comment (dlongley) [inline main.js:140]:
```suggestion
      '"credentialStatus" with type "StatusList2021Entry" and status purpose ' +
      `"${statusPurpose}" not found.`);
```

File: main.js (hunk 13 above)
Comment (dlongley) [reply main.js:133]: I think this error could be confusing when there are other valid status types ... so I suggest we combine its text with the error below.

File: main.js [hunk 15]
```diff
@@ -340,18 +338,21 @@ function _isObject({credential}) {
...
  *
- * @returns {Array<object>} An array of statuses.
+ * @returns {Array<object>} An array of statuses with type
+ *   "StatusList2021Entry" or an empty array if there are no matching types.
  */
 function _getStatuses({credential}) {
   const {credentialStatus} = credential;
-  if(Array.isArray(credentialStatus)) {
-    return credentialStatus;
+  if(Array.isArray(credentialStatus) && credentialStatus.length > 0) {
```
Comment (dlongley) [inline main.js:346]:
```suggestion
  if(Array.isArray(credentialStatus)) {
```

**Review by dlongley — APPROVED** [review]
Approving but with a few suggestions to apply before I can merge it. Thanks!

File: main.js (hunk 11 above)
Comment (JSAssassin) [reply main.js]: Addressed [7458900](https://github.com/digitalbazaar/vc-status-list/pull/6/commits/745890086265f6f4e544e933399aba9dbe78e6cb)

File: main.js (hunk 12 above)
Comment (JSAssassin) [reply main.js]: Updated.

File: main.js (hunk 13 above)
Comment (JSAssassin) [reply main.js]: Updated [7458900](https://github.com/digitalbazaar/vc-status-list/pull/6/commits/745890086265f6f4e544e933399aba9dbe78e6cb)

File: main.js (hunk 14 above)
Comment (JSAssassin) [reply main.js]: Updated.

**Comment by JSAssassin** [comment]
@dlongley I've addressed all the suggestions. I think this is ready to be merged and released.

**Review by dlongley — APPROVED** [review]
